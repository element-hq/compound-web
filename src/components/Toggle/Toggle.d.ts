import React, { PropsWithChildren } from "react";
type ToggleProps = {
    className?: string;
    onMouseDown?: (e: React.MouseEvent<HTMLInputElement, MouseEvent>) => void;
} & React.ComponentPropsWithoutRef<"input">;
/**
 * Standalone toggle component to be used with a Radix form control
 * See https://www.radix-ui.com/docs/primitives/components/form#composing-with-your-own-components
 */
export declare const Toggle: ({ className, onMouseDown, ...props }: PropsWithChildren<ToggleProps>) => React.JSX.Element;
export {};
//# sourceMappingURL=Toggle.d.ts.map