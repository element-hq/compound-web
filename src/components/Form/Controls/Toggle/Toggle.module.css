/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

.container {
  display: grid;
  inline-size: 2.25rem;
  block-size: 1.25rem;
  box-sizing: border-box;
}

.container > * {
  grid-area: 1/1;
  inline-size: inherit;
  block-size: inherit;
  box-sizing: inherit;
}

.container > .input {
  opacity: 0;
  margin: 0;
  cursor: pointer;
}

.container > .input[disabled] {
  cursor: not-allowed;
}

.ui {
  pointer-events: none;
  border-radius: var(--cpd-radius-pill-effect);
  border: 1px solid var(--cpd-color-border-interactive-primary);
  background: var(--cpd-color-bg-canvas-default);
  position: relative;
  padding: 1px;
  transition-duration: 0.2s;
  transition-timing-function: ease-in-out;
  transition-property: background-color, border-color;
}

.input:checked + .ui {
  background: var(--cpd-color-bg-accent-rest);
  border-color: var(--cpd-color-bg-accent-rest);
}

.ui::after {
  content: "";
  display: block;
  block-size: 100%;
  aspect-ratio: 1 / 1;
  border-radius: 50%;
  background: var(--cpd-color-icon-secondary);
  transform: translateX(0);
  transition-duration: 0.2s;
  transition-timing-function: ease-in-out;
  transition-property: background-color, transform;
}

/* Note the use of :focus-visible rather than :focus to avoid showing the focus
ring after a simple click */
.input:focus-visible + .ui {
  outline: 2px solid var(--cpd-color-border-focused);
  outline-offset: 1px;
}

:checked + .ui::after {
  transform: translateX(100%);
  background: var(--cpd-color-icon-on-solid-primary);
}

@media (hover) {
  :checked:not([disabled]):hover + .ui {
    background: var(--cpd-color-bg-accent-hovered);
    border-color: var(--cpd-color-bg-accent-hovered);
  }
}

:checked:active + .ui {
  background: var(--cpd-color-bg-accent-hovered);
  border-color: var(--cpd-color-bg-accent-hovered);
}

.input[readonly] {
  pointer-events: none;
}

.input[readonly] + .ui {
  border-color: var(--cpd-color-border-interactive-secondary);
  background: var(--cpd-color-bg-subtle-secondary);
}

.input[readonly] + .ui::after {
  background: var(--cpd-color-icon-secondary);
}

.input[disabled] + .ui {
  border-color: var(--cpd-color-border-disabled);
  background: var(--cpd-color-bg-canvas-disabled);
}

.input[disabled] + .ui::after {
  background: var(--cpd-color-bg-action-primary-disabled);
}

.input[readonly]:checked + .ui {
  border-color: var(--cpd-color-icon-secondary);
  background: var(--cpd-color-icon-secondary);
}

.input[readonly]:checked + .ui::after {
  background: var(--cpd-color-icon-on-solid-primary);
}

.input[disabled]:checked + .ui {
  background: var(--cpd-color-bg-action-primary-disabled);
  border-color: var(--cpd-color-bg-action-primary-disabled);
}

.input[disabled]:checked + .ui::after {
  background: var(--cpd-color-icon-on-solid-primary);
}

@media (hover) {
  .input:not(:checked, [disabled], [readonly]):hover + .ui {
    border-color: var(--cpd-color-border-interactive-hovered);

    /** TODO: have the shadow in the design tokens */
    box-shadow: 0 1.2px 2.4px 0 rgb(0 0 0 / 15%);
  }
}

.input:not(:checked, [disabled], [readonly]):active + .ui {
  border-color: var(--cpd-color-border-interactive-hovered);
}
