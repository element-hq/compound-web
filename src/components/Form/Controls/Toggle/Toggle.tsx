/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import classnames from "classnames";
import React, {
  ComponentProps,
  ComponentRef,
  forwardRef,
  PropsWithChildren,
} from "react";
import styles from "./Toggle.module.css";
import { Control } from "@radix-ui/react-form";

type ToggleProps = {
  /**
   * The CSS class name.
   */
  className?: string;
} & Omit<ComponentProps<"input">, "type">;

/**
 * A toggle component.
 */
export const ToggleInput = forwardRef<
  HTMLInputElement,
  PropsWithChildren<ToggleProps>
>(function Toggle({ className, ...props }, ref) {
  const classes = classnames(styles.container, className);
  return (
    <div className={classes}>
      <input
        role="switch"
        ref={ref}
        className={styles.input}
        {...props}
        type="checkbox"
      />
      <div className={styles.ui} />
    </div>
  );
});

/**
 * A styled checkbox input wrapped in a `Control` component, for use in Radix forms.
 */
export const ToggleControl = forwardRef<
  ComponentRef<typeof ToggleInput>,
  ComponentProps<typeof ToggleInput>
>(function ToggleControl(props, ref) {
  return (
    <Control asChild>
      <ToggleInput ref={ref} {...props} />
    </Control>
  );
});
