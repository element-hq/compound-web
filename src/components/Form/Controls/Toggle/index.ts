/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

export { ToggleControl, ToggleInput } from "./Toggle";
