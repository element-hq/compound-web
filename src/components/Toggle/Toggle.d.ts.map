{"version":3,"file":"Toggle.d.ts","sourceRoot":"","sources":["../../../../src/components/Toggle/Toggle.tsx"],"names":[],"mappings":"AAiBA,OAAO,KAAK,EAAE,EAAE,iBAAiB,EAAE,MAAM,OAAO,CAAC;AAGjD,KAAK,WAAW,GAAG;IACjB,SAAS,CAAC,EAAE,MAAM,CAAC;IACnB,WAAW,CAAC,EAAE,CAAC,CAAC,EAAE,KAAK,CAAC,UAAU,CAAC,gBAAgB,EAAE,UAAU,CAAC,KAAK,IAAI,CAAC;CAC3E,GAAG,KAAK,CAAC,wBAAwB,CAAC,OAAO,CAAC,CAAC;AAE5C;;;GAGG;AACH,eAAO,MAAM,MAAM,yCAIhB,kBAAkB,WAAW,CAAC,sBAiBhC,CAAC"}