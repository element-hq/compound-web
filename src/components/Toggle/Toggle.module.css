/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.toggle {
  display: grid;
  width: 2.75rem;
  height: 1.5rem;
  box-sizing: border-box;
}

.toggle > * {
  grid-area: 1/1;
  width: inherit;
  height: inherit;
  box-sizing: inherit;
}

.toggle > input[type="checkbox"] {
  opacity: 0;
  margin: 0;
  cursor: pointer;
}

.toggle > input[type="checkbox"][disabled] {
  cursor: not-allowed;
}

.toggle-ui {
  pointer-events: none;
  border-radius: var(--cpd-radius-pill-effect);
  border: 1px solid var(--cpd-color-border-interactive-primary);
  background: var(--cpd-color-bg-canvas-default);
  position: relative;
  padding: 1px;
  transition: all 0.2s ease;
}

:checked + .toggle-ui {
  background: var(--cpd-color-bg-action-primary-rest);
  border-color: var(--cpd-color-bg-action-primary-rest);
}

[type="checkbox"]:hover + .toggle-ui {
  border-color: var(--cpd-color-border-interactive-hovered);
}

:checked:hover + .toggle-ui {
  background: var(--cpd-color-bg-action-primary-hovered);
  border-color: var(--cpd-color-bg-action-primary-hovered);
}

.toggle-ui::after {
  content: "";
  display: block;
  height: 100%;
  aspect-ratio: 1 / 1;
  border-radius: 50%;
  background: var(--cpd-color-bg-action-primary-rest);
  transform: translateX(0);
  transition: all 0.2s ease-in-out;
}

[type="checkbox"]:focus + .toggle-ui {
  outline: 2px solid var(--cpd-color-border-focused);
  outline-offset: 1px;
}

:checked + .toggle-ui::after {
  transform: translateX(100%);
  background: var(--cpd-color-icon-on-solid-primary);
}

[type="checkbox"][disabled] + .toggle-ui {
  border-color: var(--cpd-color-border-disabled);
  background: var(--cpd-color-bg-canvas-disabled);
}

[type="checkbox"][disabled] + .toggle-ui::after {
  background: var(--cpd-color-bg-action-primary-disabled);
}

[type="checkbox"][disabled]:checked + .toggle-ui {
  background: var(--cpd-color-bg-action-primary-disabled);
  border-color: var(--cpd-color-bg-action-primary-disabled);
}

[type="checkbox"][disabled]:checked + .toggle-ui::after {
  background: var(--cpd-color-icon-on-solid-primary);
}
