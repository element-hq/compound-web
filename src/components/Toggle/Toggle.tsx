/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, { forwardRef, PropsWithChildren } from "react";
import styles from "./Toggle.module.css";

type ToggleProps = {
  className?: string;
  onMouseDown?: (e: React.MouseEvent<HTMLInputElement, MouseEvent>) => void;
} & React.ComponentPropsWithoutRef<"input">;

/**
 * Standalone toggle component to be used with a Radix form control
 * See https://www.radix-ui.com/docs/primitives/components/form#composing-with-your-own-components
 */
export const Toggle = forwardRef<
  HTMLInputElement,
  PropsWithChildren<ToggleProps>
>(function Toggle({ className, onMouseDown, ...props }, ref) {
  const classes = classnames(styles.toggle, className);
  return (
    <div className={classes}>
      <input
        ref={ref}
        {...props}
        type="checkbox"
        onMouseDown={(e) => {
          // Prevents the `focus` event from being fired when clicked
          // but still reliably works when navigating via the keyboard
          e.preventDefault();
          onMouseDown?.(e);
        }}
      />
      <div className={styles["toggle-ui"]} />
    </div>
  );
});
