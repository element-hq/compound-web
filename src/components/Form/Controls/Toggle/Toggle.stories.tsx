/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { Meta, StoryObj } from "@storybook/react";

import { ToggleInput } from "./Toggle";

const meta = {
  title: "Form/Controls/Toggle",
  component: ToggleInput,
  tags: ["autodocs"],
  argTypes: {},
  args: {},
} satisfies Meta<typeof ToggleInput>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Unchecked: Story = {
  args: {
    defaultChecked: false,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=119-249",
    },
  },
};

export const Checked: Story = {
  args: {
    defaultChecked: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=119-252",
    },
  },
};

export const DisabledUnchecked: Story = {
  args: {
    disabled: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=119-261",
    },
  },
};

export const DisabledChecked: Story = {
  args: {
    disabled: true,
    defaultChecked: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=119-264",
    },
  },
};

export const ReadonlyUnchecked: Story = {
  args: {
    readOnly: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=2457-26212",
    },
  },
};

export const ReadonlyChecked: Story = {
  args: {
    readOnly: true,
    defaultChecked: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=2457-26215",
    },
  },
};

export const Focused: Story = {
  args: {
    autoFocus: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=119-267",
    },
  },
};
