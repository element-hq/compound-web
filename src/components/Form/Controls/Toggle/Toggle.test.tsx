/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import React from "react";

import { ToggleInput } from "./Toggle";

describe("Toggle", () => {
  it("renders", () => {
    const { asFragment } = render(<ToggleInput />);
    expect(asFragment()).toMatchSnapshot();
  });
});
