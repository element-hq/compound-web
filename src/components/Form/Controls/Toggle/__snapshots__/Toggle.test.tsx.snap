// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle > renders 1`] = `
<DocumentFragment>
  <div
    class="_container_d9763f"
  >
    <input
      class="_input_d9763f"
      role="switch"
      type="checkbox"
    />
    <div
      class="_ui_d9763f"
    />
  </div>
</DocumentFragment>
`;
